src/components/select/select-list/select-list.component.js

```diff
@@ -93,63 +107,92 @@ const SelectList = React.forwardRef(
+
+    const childrenList = useMemo(() => React.Children.toArray(children), [
       children,
     ]);
 
-    const childrenWithListProps = useMemo(
-      () =>
-        React.Children.map(children, (child, index) => {
-          if (!child || (child.type !== Option && child.type !== OptionRow)) {
-            return child;
-          }
+    const optionChildrenList = useMemo(
+      () => childrenList.filter((child) => child.type === Option),
+      [childrenList]
+    );
 
-          const newProps = {
-            index,
-            id: childIds[index],
-            onSelect: handleSelect,
-            hidden: isLoading && React.Children.count(children) === 1,
-            ref: optionRefList[index],
-          };
+    const { measureElement } = virtualizer;
 
-          return React.cloneElement(child, newProps);
-        }),
-      [children, handleSelect, isLoading, optionRefList, childIds]
-    );
+    const measureCallback = (element) => {
+      // need a guard to prevent crash with too many rerenders when closing the list
+      if (isOpen) {
+        measureElement(element);
+      }
+    };
 
-    const childrenList = useMemo(
-      () => React.Children.toArray(childrenWithListProps),
-      [childrenWithListProps]
-    );
+    const renderedChildren = items.map((item) => {
+      const { index, start } = item;
+      const child = childrenList[index];
 
-    const lastOptionIndex = useMemo(() => {
-      let lastIndex = 0;
+      if (!child) {
+        return child;
+      }
 
-      childrenList.forEach((element, index) => {
-        if (element.type === Option || element.type === OptionRow) {
-          lastIndex = index;
-        }
-      });
+      const optionChildIndex = optionChildrenList.indexOf(child);
+      const isOption = optionChildIndex > -1;
+
+      const newProps = {
+        index,
+        id: childIds[index],
+        onSelect: handleSelect,
+        hidden: isLoading && childrenList.length === 1,
+        // these need to be inline styles rather than implemented in styled-components to avoid it generating thousands of classes
+        style: {
+          transform: `translateY(${start}px)`,
+        },
+        "aria-setsize": isOption ? optionChildrenList.length : undefined,
+        "aria-posinset": isOption ? optionChildIndex + 1 : undefined,
+        // needed to dynamically compute the size
+        ref: measureCallback,
+        "data-index": index,
+      };
+
+      return React.cloneElement(child, newProps);
+    });
 
-      return lastIndex;
+    const lastOptionIndex = useMemo(() => {
+      // findLastIndex does not exist in Node till v18, and is only in browsers since the start of 2022
+      // so we can't really use it. This should be updated when there is better support!
+      for (let i = childrenList.length - 1; i >= 0; i--) {
+        const child = childrenList[i];
+        if (child.type === Option || child.type === OptionRow) {
+          return i;
+        }
+      }
+      return -1;
```

Comment:
I'd go with this if you refactor as it's more readable than what I have above

Reply:
in the end I've adopted a hybrid of the 2 approaches. The linter doesn't like using `--` but it seems there's an exception for the header of a `for` loop. So I've kept this as a `for` loop and mostly similar to what I had, but I have adopted your suggestion of having a single return and just `break`ing the loop rather than having an early return when the correct child is found.